File: Leetcode/Python3/3. Longest Substring Without Repeating Characters/solution.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/

# Ref: https://leetcode.com/problems/longest-substring-without-repeating-characters/discuss/1731/A-Python-solution-85ms-O(n)
from operator import index


class Solution:
    def lengthOfLongestSubstring(self, s):
        start = maxLength = 0
        usedChar = {}

        for i in range(len(s)):
            # 현재문자가 이미 사용된적있다면 그 문자가 마지막으로나온 인덱스 + 1을 start지점으로 지정한다
            # start가 뒤로 가는것을 방지하기 위해 start <= usedChar[s[i]]를 체크한다
            if s[i] in usedChar and start <= usedChar[s[i]]:
                start = usedChar[s[i]] + 1
            else:
                # maximum length can only be increased when a new letter is appear 
                maxLength = max(maxLength, i - start + 1)

            usedChar[s[i]] = i

        return maxLength

File: Leetcode/C++/3. Longest Substring Without Repeating Characters/solution.cpp
// https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
// Time complexity: O(N)

#include <map>
#include <string>

using namespace std;

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, int> used;
        // left = start index of substring
        int left = 0, maxLength = 0;
        for (int idx = 0; idx < s.size(); idx++)
        {
            if (used.find(s[idx]) == used.end())
            {
                used.insert({s[idx], idx});
            }
            // If current char is already used update index
            else
            {
                // Prevent left index move left
                if (left < used[s[idx]] + 1)
                {
                    left = used[s[idx]] + 1;
                }
                used[s[idx]] = idx;
            }
            // Compare maxLength and current substring
            maxLength = max(maxLength, idx - left + 1);
        }
        return maxLength;
    }
};